cluster-autoscaler/processors/nodegroupset/balancing_processor.go

```diff
@@ -179,5 +193,66 @@ func (b *BalancingNodeGroupSetProcessor) BalanceScaleUpBetweenGroups(context *co
 	return result, nil
 }
 
+// averageNodegroupAllocatableResources collects the average capacity by nodegroup id,
+// so we can add new nodes to the one with the lowest capacity instead of the one with the lowest node count
+func (b *BalancingNodeGroupSetProcessor) averageNodegroupAllocatableResources(context *context.AutoscalingContext, groups []cloudprovider.NodeGroup, nodes []*apiv1.Node) map[string]int {
+	// collect info
+	averageNodegroupCapacities := map[string]int{}
+	nodegroupCounts := map[string]int{}
+	fallback := 0
+
+	if context.BalanceSimilarNodeGroupsBy == BalanceByCpu || context.BalanceSimilarNodeGroupsBy == BalanceByMemory {
+		for _, ng := range groups {
+			averageNodegroupCapacities[ng.Id()] = 0
+			nodegroupCounts[ng.Id()] = 0
+			ngNodes, err := ng.Nodes()
+			if err != nil {
+				klog.V(1).Infof("Failed to get nodes from ASG %v: %v", ng.Id(), err)
```

Comment:
nit: s/ASG/NodeGroup/, since ASG is a AWS-specific context and this code is shared between different cloud providers.